### ISSUE TITLE
UnboundLocalError: cannot access local variable 'x' where it is not associated with a value

### ISSUE THREAD

**Issue by kristiker** [pytest-dev/pytest#11028]
There seems to be a regression in pytest version `7.3.x` when a **walrus** operator is used in an assert line.
Code:

```py
import json
import pytest

def test_json_encoder():
  assert (object:="foo") in json.dumps(object)
```

Fails the test with error:
```shell
UnboundLocalError: cannot access local variable 'object' where it is not associated with a value
```

in pytest version `7.3.x`, whereas with pytest version `7.2.x` it passes successfully. My Python version is `3.11`.

Looks like it has to do with PR #10758.

**Comment by The-Compiler** [comment]
Yep, I could indeed bisect this to #10758 / 6e478b094787f3d2bf2db3676d330c7a13ac6b98.

cc @aless10

**Comment by aless10** [comment]
> Yep, I could indeed bisect this to #10758 / [6e478b0](https://github.com/pytest-dev/pytest/commit/6e478b094787f3d2bf2db3676d330c7a13ac6b98).
> 
> cc @aless10

Hi, I think this is definitely something I worked on. It seems that I did not cover this case. @The-Compiler I can work on this one by the end of the week it's ok with you.

**Comment by The-Compiler** [comment]
Sure thing, feel free and thanks for taking a look!
